src/java.base/share/classes/sun/security/jca/ProviderList.java

```diff
@@ -146,6 +154,17 @@ public static ProviderList remove(ProviderList providerList, String name) {
     // Create a new ProviderList from the specified Providers.
     // This method is for use by SunJSSE.
     public static ProviderList newList(Provider ... providers) {
+        if (RestrictedSecurity.isEnabled()) {
+            List<Provider> allowedProviders = new ArrayList<>(Arrays.asList(providers));
+            for (Provider p : providers) {
+                if (!RestrictedSecurity.isProviderAllowed(p.getName())) {
+                    // We're in restricted security mode which does not allow this provider,
+                    // remove it from the list.
+                    allowedProviders.remove(p);
+                }
+            }
```

Comment:
(I thought I made this comment earlier) There's no need to put providers in the list unless they're allowed, avoiding the cost of finding those that need to be removed.
```java
            List<Provider> allowedProviders = new ArrayList<>();
            for (Provider p : providers) {
                if (RestrictedSecurity.isProviderAllowed(p.getName())) {
                    // This provider is allowed, add it the list.
                    allowedProviders.add(p);
                }
            }
```